netwayste/src/utils/ping.rs

```diff
@@ -0,0 +1,189 @@
+use rand::random;
+
+/// This indicates the number of samples needed by the latency filter to calculate a statistically
+/// meaningful average.
+const LATENCY_FILTER_DEPTH: usize = 12;
+
+#[derive(Serialize, Deserialize, Clone, PartialEq, Eq, Debug)]
+pub struct PingPong {
+    pub nonce: u64
+}
+
+impl PingPong {
+    pub fn ping() -> PingPong {
+        PingPong {
+            nonce: random::<u64>()
+        }
+    }
+
+    pub fn pong(nonce: u64) -> PingPong {
+        PingPong {
+            nonce
+        }
+    }
+}
+
+/// A moving-average filter used to level out the latencies calculated from network request/response times.
+#[derive(PartialEq,Eq,Debug, Clone)]
+pub struct LatencyFilter {
+    pub average_latency_ms: Option<u64>,
+    running_sum: u64,
+    history: VecDeque<u64>,
+    start_timestamp: Instant,
+    in_progress: bool,
+}
+
+impl LatencyFilter {
+    pub fn new() -> LatencyFilter {
+        LatencyFilter {
+            average_latency_ms: None,
+            running_sum: 0,
+            history: VecDeque::with_capacity(LATENCY_FILTER_DEPTH),
+            start_timestamp: Instant::now(),
+            in_progress: false,
+        }
+    }
+
+    #[deny(unused_variables)]
+    pub fn reset(&mut self) {
+        let Self {
+            ref mut average_latency_ms,
+            ref mut running_sum,
+            ref mut history,
+            ref mut start_timestamp,
+            ref mut in_progress,
+        } = *self;
+
+        *average_latency_ms = None;
+        *running_sum = 0;
+        history.clear();
+        *start_timestamp = Instant::now();
+        *in_progress = false;
+    }
+
+    pub fn start(&mut self) {
+        self.start_timestamp = Instant::now();
+        self.in_progress = true;
+    }
+
+    pub fn update(&mut self) {
+        if !self.in_progress {
+            error!("The LatencyFilter's start() was not called so a duration cannot be computed.");
+            let elapsed = self.start_timestamp.elapsed();
+            error!(
+                "LatencyFilter.start_timestamp snapshot was {}.{}.{} seconds ago.",
+                elapsed.as_secs(),
+                elapsed.as_millis(),
+                elapsed.as_micros()
+            );
+        }
+
+        let latency = Instant::now().duration_since(self.start_timestamp);
+        let latency_ms = latency.as_millis() as u64;
+
+        self.running_sum += latency_ms;
+        self.history.push_back(latency_ms);
+
+        // Wait for the filter to be populated
+        if self.history.len() > LATENCY_FILTER_DEPTH {
+            // unwraps safe b/c of length check
+            let oldest = self.history.pop_front().unwrap();
+            self.running_sum -= oldest;
+
+            // average round-trip time
+            let average_rtt_ms = (self.running_sum as f64 / LATENCY_FILTER_DEPTH as f64) as u64;
+            let average_latency_ms = average_rtt_ms / 2;
+
+            // TODO: Remove me once the server list is implemented in the UI
```

Comment:
Why not `PR_GATE` this so we don't have a bunch of log output after the PR is merged?

Reply:
If you want to see feedback maybe it would be better to have cli-client read the `average_latency_ms`.

Reply:
Check out the latest commit :smile: .

Also `PR_GATE`s are my way of keeping track things that must be resolved before the PR can be approve and merged.